Automattic-Tracks-iOS/Event Logging/LogEncryptor.swift

```diff
@@ -0,0 +1,104 @@
+import Foundation
+import Sodium
+
+class LogEncryptor {
+
+    private let chunkSize = 4096
+    private let publicKey: Bytes
+
+    enum LogEncryptorError: Error {
+        case unableToReadFile
+        case unableToEncryptFile
+        case logSecretTooLong
+        case unableToWriteFile
+    }
+
+    init(withPublicKey key: Bytes) {
+        self.publicKey = key
+    }
+
+    func encryptLog(_ log: LogFile) throws -> URL {
+
+        /// Set up our output path
+        let uniqueSegment = UUID().uuidString + "-" + log.url.lastPathComponent
+        let outputURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(uniqueSegment)
+        assert(outputURL.path != log.url.path, "The output file path must be unique")
+
+        /// Do the encrypted stream setup
+        let sodium = Sodium()
+        let secretkey = sodium.secretStream.xchacha20poly1305.key()
+        let encryptedKey = try encryptSecretWithSodium(secret: secretkey)
+        let stream_enc = sodium.secretStream.xchacha20poly1305.initPush(secretKey: secretkey)!
+        let header = stream_enc.header()
+
+        /// Prep our file handles
+        try initializeOutputFile(at: outputURL)
+        let inputFileHandle = try FileHandle(forReadingFrom: log.url)
+        let outputFileHandle = try FileHandle(forWritingTo: outputURL)
+
+        defer {
+            inputFileHandle.closeFile()
+            outputFileHandle.closeFile()
+        }
+
+        /// Write JSON Preamble
+        outputFileHandle.write("""
+        {
+            "keyedWith": "v1",
+            "encryptedKey": "\(Data(encryptedKey).base64EncodedString())",
+            "header": "\(Data(header).base64EncodedString())",
+            "uuid": "\(log.uuid)",
+            "messages": [\n
+        """.data(using: .utf8)!)
+
+        /// Write the encrypted file in chunks as JSON
+        var shouldContinue = true
+        repeat {
+
+            let data = inputFileHandle.readData(ofLength: chunkSize)                   // read the data
+            guard let message = stream_enc.push(message: [UInt8](data)) else {         // encrypt the data
+                throw LogEncryptorError.unableToEncryptFile
+            }
+            try writeTo(outputFileHandle, message: message)                            // write the data
+
+            shouldContinue = data.count == chunkSize
+        } while shouldContinue
```

Comment:
Oguz and I wondered whether we should build a `struct` and convert it to JSON instead of manually writing the JSON. We agreed that what we have here is probably better since we could save active memory usage.

I'd like to suggest that we extract the complexity of the loop to a separate method though. This should make this block focused on _writing only_ and should probably make it easier to reason about. 

We can do this by adding a method for `FileHandle`:

```swift
private extension FileHandle {
    /// Loop through every chunk until EOF
    ///
    /// Calls `next` for every chunk.
    func readDataUntilEOF(chunkSize: Int, next: ((Data) throws -> Void)) throws {
        // TODO Do the loop here
        repeat {
            let data = readData(ofLength: chunkSize)
            try next(data)
        } while ...
    }
}
```

The method name can be better. 😄 

And then, this block can call that like so:

```swift
        try inputFileHandle.readDataUntilEOF(chunkSize: chunkSize) { data in
            guard let message = stream_enc.push(message: [UInt8](data)) else {         // encrypt the data
                throw LogEncryptorError.unableToEncryptFile
            }
            // Can message contain double quotes? Or escape JSON characters?
            try writeTo(outputFileHandle, message: message)                            // write the data
        }
```

Reply:
Yep! This makes sense – done in 558b7882ca2965efea40cc829bc85be3985305f1. I used `readChunkedDataToEndOfFile`, based on the pre-existing `readDataToEndOfFile` method.

Reply:
Looks nice! 🎉